### DISCUSSION TITLE
Oracle standard module

### DISCUSSION THREAD

**Discussion by dpc** [fedimint/fedimint#3620] [Fedimint Modules]
For UX purposes it's useful if the user can know e.g. fiat prices.

To keep it federated, it would be great if Fedimind could run it.

We could add a new core module where the whole purpose is agreeing on certain facts (e.g. BTC/USD price).

#### Voting

Voting would happen exactly like it happens for block fee rate and block height in wallet module already. Each peer, when they change their vote, sends a `ConsensusItem::Vote`:

```
struct Vote {
  id: u64,
  value: u64,
}
```

As you can see `index` allows voting on arbitrary number of things.

For each `id` and  peer in the Federation, the `values` would be tracked and the final value would be the median.

To prevent unconstrained blowing up disk & memory use space, each `id` needs to be enabled first with a `ConsensItem::Enable(u64)`, until threshold is reached. Before then `OracleVote`s for it are illegal. A `ConsensusItem::Disable` seems useful to have as well.

Fedimint the Open Source Project would track `id` to "underlying meaning" allocation (e.g. national currencies), with a range dedicated to custom usage.

#### Source configuration

For each `id`, each guardian should be able to configure one or more multiple sources:

* `idx` - index of the source
* `SourceType` - from a list of implemented source typs (eg. json call; jsonrpc call)
* `SourceUrl` - URL specific for the type (eg. 
* `SourceSelector` - eg. jq-like syntax to pick the value from the source
* `decimalShift` - before rounding down an integer, multiply by `10 ^ {decimalShift}`
* `SuccessPeriod` - time to query the source again after a success
* `FailurePeriod` - time to query the source again after a failure

Fedimint should come with the defaults (probably randomly chosen from at least 3) for each standard `id`

Note: this is a peer-local configuration, so each guardian is free to change it anytime.

### Source Modes

How to query multiple sources for same `id` with different indexes:

* `failover` - from lowest to hightest, use first success
* `median` - query all, calculculate a median

### Overrides

For each `id` guardian should be able to specify a manual override:

* `id : u64` - which `id` to override
* `value : u64` - value to override with
* `until: Option<u64>`- unix timestamp till which the override should apply; forever if `None`

This should be possible for the `fedimint-cli` and Admin UI. 

#### Global, public

* `get_value(id: u64) -> Option<u64>`

#### Admin-Private:

* `enableId(id: u64)`
* `disableId(id: u64)`
* `setOverride(id: u64, value: u64, until: Option<u64>)`
* `setSource(idx: u32, type: String, url: String, selector: String, decimalShift: s8, sPeriod: u32, fPeriod: u32);`
* `getVote(idx: u64) -> u64;`
* `setSourceMode(mode: String);`

**Comment by elsirion** [comment]
I wonder if it would be useful to have two kinds of modules: oracles and the current modules. Then e.g. the block height could be an oracle module that both the LN and Wallet module depend on.

**Comment by dpc** [comment]
> Then e.g. the block height could be an oracle module that both the LN and Wallet module depend on.

Crossed my mind to make this module responsible for block height and fee rate as well.
